File: src/main/java/com/mateusz/rest/webservices/restful_web_services/jpa/UserJpaResource.java
package com.mateusz.rest.webservices.restful_web_services.jpa;

import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.servlet.support.ServletUriComponentsBuilder;

import com.mateusz.rest.webservices.restful_web_services.users.Post;
import com.mateusz.rest.webservices.restful_web_services.users.PostNotFoundException;
import com.mateusz.rest.webservices.restful_web_services.users.User;
import com.mateusz.rest.webservices.restful_web_services.users.UserNotFoundException;

import jakarta.validation.Valid;

import static org.springframework.hateoas.server.mvc.WebMvcLinkBuilder.*;

import java.net.URI;
import java.util.List;
import java.util.Optional;

import org.springframework.hateoas.EntityModel;
import org.springframework.hateoas.server.mvc.WebMvcLinkBuilder;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.DeleteMapping;
// import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;

@RestController
public class UserJpaResource {
    // private UserDaoService userDaoService;

    private UserRepository userRepository;

    private PostRepository postRepository;

    // @Autowired
    public UserJpaResource(UserRepository userRepository,
            PostRepository postRepository) {
        this.userRepository = userRepository;
        this.postRepository = postRepository;
    }

    // GET /users
    @GetMapping("/users")
    public List<User> retrieveAllUsers() {
        return userRepository.findAll();
    }

    // http://localhost:8080/users

    // EntityModel
    // WebMvcLinkBuilder

    // GET /users/{id}
    @GetMapping("/users/{id}")
    public EntityModel<User> retrieveOneUser(@PathVariable int id) {
        Optional<User> user = userRepository.findById(id);

        if (user.isEmpty()) {
            throw new UserNotFoundException("id:" + id);
        }
        EntityModel<User> entityModel = EntityModel.of(user.get());

        WebMvcLinkBuilder link = linkTo(methodOn(this.getClass()).retrieveAllUsers());

        entityModel.add(link.withRel("all-users"));

        return entityModel;
    }

    // POST /users
    @PostMapping("/users")
    public ResponseEntity<User> createOneUser(@Valid @RequestBody User user) {
        User savedUser = userRepository.save(user);
        // /users/4 => /users/{id}, user.getId
        URI location = ServletUriComponentsBuilder.fromCurrentRequest().path("/{id}")
                .buildAndExpand(savedUser.getId()).toUri();
        return ResponseEntity.created(location).build();
    }

    @DeleteMapping("/users/{id}")
    public void deleteOneUser(@PathVariable int id) {
        userRepository.deleteById(id);
    }

    @GetMapping("/users/{id}/posts")
    public List<Post> retrieveAllPostsForUser(@PathVariable int id) {
        Optional<User> user = userRepository.findById(id);

        if (user.isEmpty())
            throw new UserNotFoundException("id:" + id);

        return user.get().getPosts();
    }

    @PostMapping("/users/{id}/posts")
    public ResponseEntity<Post> createOnePostForUser(@PathVariable int id, @Valid @RequestBody Post post) {
        Optional<User> user = userRepository.findById(id);

        if (user.isEmpty())
            throw new UserNotFoundException("id:" + id);

        post.setUser(user.get());// get because it's Optional

        Post savedPost = postRepository.save(post);

        URI location = ServletUriComponentsBuilder.fromCurrentRequest().path("/{id}")
                .buildAndExpand(savedPost.getId()).toUri();
        return ResponseEntity.created(location).build();

    }

    @GetMapping("/users/{userId}/posts/{postId}")
    public EntityModel<Post> retrieveOnePost(@PathVariable int userId, @PathVariable int postId) {
        Optional<User> user = userRepository.findById(userId);

        if (user.isEmpty()) {
            throw new UserNotFoundException("id:" + userId);
        }

        Post post = postRepository.findById(postId)
                .orElseThrow(() -> new PostNotFoundException("id:" + postId));

        EntityModel<Post> entityModel = EntityModel.of(post);

        WebMvcLinkBuilder linkToUser = linkTo(methodOn(this.getClass()).retrieveOneUser(userId));
        WebMvcLinkBuilder linkToPosts = linkTo(methodOn(this.getClass()).retrieveAllPostsForUser(userId));

        entityModel.add(linkToUser.withRel("user"));
        entityModel.add(linkToPosts.withRel("all-posts"));

        return entityModel;
    }

}

File: src/main/kotlin/com/mateusz/rest/webservices/restful_web_services/spring_boot_social_media_restful_api/jpa/UserJpaResource.kt
package com.mateusz.rest.webservices.restful_web_services.spring_boot_social_media_restful_api.jpa

import com.mateusz.rest.webservices.restful_web_services.spring_boot_social_media_restful_api.users.Post
import com.mateusz.rest.webservices.restful_web_services.spring_boot_social_media_restful_api.users.PostNotFoundException
import com.mateusz.rest.webservices.restful_web_services.spring_boot_social_media_restful_api.users.User
import com.mateusz.rest.webservices.restful_web_services.spring_boot_social_media_restful_api.users.UserNotFoundException
import jakarta.validation.Valid
import org.springframework.hateoas.EntityModel
import org.springframework.hateoas.server.mvc.WebMvcLinkBuilder
import org.springframework.hateoas.server.mvc.WebMvcLinkBuilder.*
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import org.springframework.web.servlet.support.ServletUriComponentsBuilder
import java.util.*


@RestController
class UserJpaResource(
    private val userRepository: UserRepository,
    private val postRepository: PostRepository
) {

    // GET /users
    @GetMapping(value = ["/users"], produces = ["application/json", "application/xml", "application/hal+json"] )
    fun retrieveAllUsers(): List<User> {
        return userRepository.findAll().filterNotNull()
    }

    // GET /users/{id}
    @GetMapping(value = ["/users/{id}"], produces = ["application/json", "application/xml", "application/hal+json"] )
    fun retrieveOneUser(@PathVariable id: Int): EntityModel<User> {
        val user = userRepository.findById(id).orElseThrow { UserNotFoundException("User not found with id: $id") }

        val entityModel = EntityModel.of(user!!)

        val link: WebMvcLinkBuilder = linkTo(methodOn(this.javaClass).retrieveAllUsers())

        entityModel.add(link.withRel("all-users"))

        return entityModel
    }

    // POST /users
    @PostMapping(value = ["/users"], consumes = ["application/json", "application/xml", "application/hal+json"], produces = ["application/json", "application/xml", "application/hal+json"] )
    fun createOneUser(@Valid @RequestBody user: User?): ResponseEntity<User> {
        val savedUser: User = userRepository.save(user!!)
        val location = ServletUriComponentsBuilder.fromCurrentRequest().path("/{id}")
            .buildAndExpand(savedUser.id).toUri()
        return ResponseEntity.created(location).build()
    }

    @DeleteMapping(value = ["/users/{id}"], produces = ["application/json", "application/xml", "application/hal+json"])
    fun deleteOneUser(@PathVariable id: Int) {
        userRepository.deleteById(id)
    }

    @GetMapping(value = ["/users/{id}/posts"], produces = ["application/json", "application/xml", "application/hal+json"] )
    fun retrieveAllPostsForUser(@PathVariable id: Int): List<Post> {
        val user = userRepository.findById(id).orElseThrow { UserNotFoundException("User not found with id: $id") }
        return user?.posts ?: emptyList()
    }

    @PostMapping(value = ["/users/{id}/posts"], consumes = ["application/json", "application/xml", "application/hal+json"] )
    fun createOnePostForUser(@PathVariable id: Int, @Valid @RequestBody post: Post): ResponseEntity<Post> {
        val user: Optional<User> = userRepository.findById(id).map { it ?: throw UserNotFoundException("id:$id") }
        post.user = user.get() // get because it's Optional

        val savedPost: Post = postRepository.save(post)

        val location = ServletUriComponentsBuilder.fromCurrentRequest().path("/{id}")
            .buildAndExpand(savedPost.id).toUri()
        return ResponseEntity.created(location).build()
    }

    @GetMapping(value = ["/users/{userId}/posts/{postId}"], produces = ["application/json", "application/xml", "application/hal+json"] )
    fun retrieveOnePost(@PathVariable userId: Int, @PathVariable postId: Int): EntityModel<Post> {
        userRepository.findById(userId).orElseThrow { UserNotFoundException("User not found with id: $userId") }
        val post = postRepository.findById(postId).orElseThrow { PostNotFoundException("Post not found with id: $postId") }

        val entityModel = EntityModel.of(post!!)

        val linkToUser: WebMvcLinkBuilder = linkTo(methodOn(this.javaClass).retrieveOneUser(userId))
        val linkToPosts: WebMvcLinkBuilder = linkTo(methodOn(this.javaClass).retrieveAllPostsForUser(userId))

        entityModel.add(linkToUser.withRel("user"))
        entityModel.add(linkToPosts.withRel("all-posts"))

        return entityModel
    }
}